### ISSUE TITLE
Multipart copying fails in buckets enforcing encryption

### ISSUE THREAD

**Issue by the-nguyen-databricks** [aws/aws-sdk-java#3001]
### Describe the bug

Multipart copying using the SDK’s [TransferManager.copy(copyObjectRequest)](https://github.com/aws/aws-sdk-java/blob/eaa726d99e90f148ee7316c9c30a154b81ebfd0f/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/transfer/TransferManager.java#L2178) method conflicts with the [documented way of setting up a bucket enforcing SSE-C encryption](https://docs.aws.amazon.com/AmazonS3/latest/userguide/ServerSideEncryptionCustomerKeys.html#ssec-require-condition-key) using a bucket policy. Specifically, the policy requires all PUT requests to have some mandatory encryption headers (in our case, header `x-amz-server-side-encryption-customer-algorithm` is required). However, the complete multipart upload request of a multipart copying process does not have the required encryption header.

Relevant methods are:
- [TransferManager.copy(copyObjectRequest, srcS3, stateChangeListener)](https://github.com/aws/aws-sdk-java/blob/eaa726d99e90f148ee7316c9c30a154b81ebfd0f/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/transfer/TransferManager.java#L2265)
- [CopyMonitor.call()](https://github.com/aws/aws-sdk-java/blob/eaa726d99e90f148ee7316c9c30a154b81ebfd0f/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/transfer/internal/CopyMonitor.java#L130C5-L130C29)
- [CompleteMultipartCopy.call()](https://github.com/aws/aws-sdk-java/blob/eaa726d99e90f148ee7316c9c30a154b81ebfd0f/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/transfer/internal/CompleteMultipartCopy.java#L73C5-L73C29)

### Expected Behavior

The SDK should set all necessary encryption headers internally for the complete request of a multipart copying process when calling the `copy` function. Specifically, in the `call` function of the `CompleteMultipartCopy` instance of a multipart copying process, the `CompleteMultipartUploadRequest` instance created for sending the complete request should set all necessary encryption headers before the actual sending.

### Current Behavior

The SDK does not have any actions to set encryption headers internally for the complete request of a multipart copying process. As a consequence, the complete request sent will be blocked by an S3 [bucket policy](https://docs.aws.amazon.com/AmazonS3/latest/userguide/example-bucket-policies.html) which enforces SSE encryption.

```
at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleErrorResponse(AmazonHttpClient.java:1879)
at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleServiceErrorResponse(AmazonHttpClient.java:1418)
at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeOneRequest(AmazonHttpClient.java:1387)
at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1157)
at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:814)
at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:781)
at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:755)
at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:715)
at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:697)
at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:561)
at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:541)
at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:5456)
at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:5403)
at com.amazonaws.services.s3.AmazonS3Client.completeMultipartUpload(AmazonS3Client.java:3630)
at com.amazonaws.services.s3.transfer.internal.CompleteMultipartCopy.call(CompleteMultipartCopy.java:85)
at com.amazonaws.services.s3.transfer.internal.CompleteMultipartCopy.call(CompleteMultipartCopy.java:38)
at java.util.concurrent.FutureTask.run(FutureTask.java:266)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
at java.lang.Thread.run(Thread.java:748)
```

### Reproduction Steps

```java
String bucket = "a-bucket-enforcing-encryption";
String key = "a-file-large-enough-to-trigger-multipart-copying.txt";
TransferManager transfers = TransferManagerBuilder.defaultTransferManager();
CopyObjectRequest copyObjectRequest = new CopyObjectRequest(bucket, key, bucket, key);
copyObjectRequest.setSourceSSECustomerKey(new SSECustomerKey(new byte[32]));
copyObjectRequest.setDestinationSSECustomerKey(new SSECustomerKey(new byte[32]));
transfers.copy(copyObjectRequest);
```

### Possible Solution

In the `call` function of the `CompleteMultipartCopy` instance of a multipart copying process, we can extract encryption information from the original COPY request and set all encryption headers which are possible to be set with the retrieved encryption information for the `CompleteMultipartUploadRequest` instance.

### AWS Java SDK version used

1.12

### JDK version used

1.8.0_292

### Operating System and version

Ubuntu 18.04.5 LTS

**Comment by debora-ito** [comment]
@the-nguyen-databricks Confirming I can reproduce the issue with TransferManager multipart copy. Buckets with the policy requiring SSE-C for all uploads will get an AccessDenied error. 

TransferManager single-part copy works fine, but I understand it's not a reasonable workaround for large files.

**Comment by debora-ito** [comment]
A fix was released in Java SDK version `1.12.511`.

Let us know if you still see any issue after the upgrade.

**Comment by the-nguyen-databricks** [comment]
@debora-ito We have verified the fix. It works. However, the fix is actually in version `1.12.512`
